tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -2,25 +2,235 @@
+      TestData.push_back(ShaderBufferPtr[i]);
+  } else {
+    DataTypeT *ShaderBufferPtr =
+        reinterpret_cast<DataTypeT *>(ShaderBuffer.data());
+    for (size_t i = 0; i < NumElements; ++i)
+      TestData.push_back(ShaderBufferPtr[i]);
+  }
+}
+
+template <typename DataTypeT>
+bool LongVector::doValuesMatch(DataTypeT A, DataTypeT B, float Tolerance,
+                               LongVector::ValidationType) {
+  if (Tolerance == 0.0f)
+    return A == B;
+
+  DataTypeT Diff = A > B ? A - B : B - A;
+  return Diff <= Tolerance;
+}
+
+bool LongVector::doValuesMatch(HLSLBool_t A, HLSLBool_t B, float,
+                               LongVector::ValidationType) {
+  return A == B;
+}
+
+bool LongVector::doValuesMatch(HLSLHalf_t A, HLSLHalf_t B, float Tolerance,
+                               LongVector::ValidationType ValidationType) {
+  switch (ValidationType) {
+  case LongVector::ValidationType_Epsilon:
+    return CompareHalfEpsilon(A.Val, B.Val, Tolerance);
+  case LongVector::ValidationType_Ulp:
+    return CompareHalfULP(A.Val, B.Val, Tolerance);
+  default:
+    WEX::Logging::Log::Error(
+        L"Invalid ValidationType. Expecting Epsilon or ULP.");
+    return false;
+  }
+}
+
+bool LongVector::doValuesMatch(float A, float B, float Tolerance,
+                               LongVector::ValidationType ValidationType) {
+  switch (ValidationType) {
+  case LongVector::ValidationType_Epsilon:
+    return CompareFloatEpsilon(A, B, Tolerance);
+  case LongVector::ValidationType_Ulp: {
+    // Tolerance is in ULPs. Convert to int for the comparison.
+    const int IntTolerance = static_cast<int>(Tolerance);
+    return CompareFloatULP(A, B, IntTolerance);
+  };
+  default:
+    WEX::Logging::Log::Error(
+        L"Invalid ValidationType. Expecting Epsilon or ULP.");
+    return false;
+  }
+}
+
+bool LongVector::doValuesMatch(double A, double B, float Tolerance,
+                               LongVector::ValidationType ValidationType) {
+  switch (ValidationType) {
+  case LongVector::ValidationType_Epsilon:
+    return CompareDoubleEpsilon(A, B, Tolerance);
+  case LongVector::ValidationType_Ulp: {
+    // Tolerance is in ULPs. Convert to int64_t for the comparison.
+    const int64_t IntTolerance = static_cast<int64_t>(Tolerance);
+    return CompareDoubleULP(A, B, IntTolerance);
+  };
+  default:
+    WEX::Logging::Log::Error(
+        L"Invalid ValidationType. Expecting Epsilon or ULP.");
+    return false;
+  }
+}
+
+template <typename DataTypeT>
+bool LongVector::doVectorsMatch(const std::vector<DataTypeT> &ActualValues,
+                                const std::vector<DataTypeT> &ExpectedValues,
+                                float Tolerance,
+                                LongVector::ValidationType ValidationType) {
+  // Stash mismatched indexes for easy failure logging later
+  std::vector<size_t> MismatchedIndexes;
+  VERIFY_IS_TRUE(ActualValues.size() == ExpectedValues.size(),
```

Comment:
This is probably good enough, but it there value in comparing the values that are present by looping over the shorter vector?